neps/nep-519-yield-execution.md

```diff
@@ -0,0 +1,260 @@
+
+## Motivation
+
+There exist some situations where when a smart contract on NEAR is called, it will only be able to provide an answer at some arbitrary time in the future.  So the callee needs a way to defer replying to the caller till this time in future.
+
+Examples include when a smart contract (`S`) provides MPC signing capabilities parties external to the NEAR protocol are computing the signature.  The rough steps are:
+
+1. Signer contract provides a function `fn sign_payload(Payload, ...)`.
+2. When called, the contract defers replying to the caller.
+3. External indexers are monitoring the transactions on the contract; they observe the new signing request, compute a signature, and call another function `fn signature_available(Signature, ...)` on the signer contract.
+4. The signer contract validates the signature and if validate, replies to the original caller.
+
+Today, the NEAR protocol has no sensible way to defer replying to the caller in step 2 above.  This proposal proposes adding two following new host functions to the NEAR protocol:
+
+- `yield`: this can be called by a contract to indicate to the protocol that it is not ready yet to reply to its caller.
+- `resume`: a contract can use this mechanism to indicate to a protocol that it is now ready to reply to a caller that it had deferred earlier.
+
+If these two host functions were available, then `yield` would be used in step 2 above and `resume` would be used in step 4 above.
+
+## Specification
+
+The proposal is to add the following host functions to the NEAR protocol:
+
+
+```rust
+/// Instructs the protocol that the smart contract is not yet ready to respond
+/// to its caller yet.  The smart contract promises to call
+/// `promise_yield_resume()` within X (TBD) blocks.  When
+/// `promise_yield_resume()` is called, the protocol will call the method on the
+/// smart contract that is identified by `method_name_len` and `method_name_ptr`
+/// and this method may or may not respond to the caller.
+///
+/// `arguments_len` and `arguments_ptr` provide an initial blob of arguments
+/// that will be passed to the method.
+///
+/// If the contract fails to call `promise_yield_resume()` within X blocks, then
+/// the protocol will call the method with a timeout error.
+///
+/// Similar to the `gas` parameter in
+/// [promise_create](https://github.com/near/nearcore/blob/a908de36ab6f75eb130447a5788007e26d05f93e/runtime/near-vm-runner/src/logic/logic.rs#L1281),
+/// the `gas` parameter is a prepayment for the gas that would be used to
+/// execute the method.
+///
+/// `gas_weight`: as specified in
+/// [this](https://github.com/near/NEPs/blob/master/neps/nep-0264.md) NEP, this
+/// improves the devX of specifying a portion of the remaining gas for executing
+/// the method instead of specifying a precise amount.
```

Comment:
We don't need to reiterate the motivation for that NEP. Better refer to `promise_create` the same way like it was done for `gas`.

Reply:
Done.